### ISSUE TITLE
clang-format warning multiple different client offset_decoding detected for buffer

### ISSUE THREAD

**Issue by JeremiahCheatham** [LunarVim/LunarVim#2597]
### Problem description

I was able to  fix formatting for c code to 4 spaces in the main config. No idea if this is right but it worked.
`~/.config/lvim/config.lua`

```local formatters = require "lvim.lsp.null-ls.formatters"
formatters.setup {
    {
        command = "clang-format",
        filetype = { "c", "cpp", "cs", "java" },
        extra_args = { "--style", "{IndentWidth: 4}" }
    }
}
```
However i suddenly got spammed with 
`warning multiple different client offset_decoding detected for buffer`

I tried this solution https://github.com/LunarVim/LunarVim/issues/2209#issuecomment-1019826987
`/usr/local/share/nvim/runtime/lua/vim/lsp/util.lua` Doesn't exist.
`/usr/share/nvim/runtime/lua/vim/lsp/util.lua` Does exist.
`offset_encoding = offset_encoding or "utf-16"` Added to line 365 had no effect at all.

However i tried something from this post https://github.com/neovim/neovim/pull/16694#issuecomment-996947306
`~/.local/share/lunarvim/lvim/lua/lvim/lsp/init.lua` this file.
`local capabilities = vim.lsp.protocol.make_client_capabilities()` Under this line 33.
 `capabilities.offsetEncoding = { "utf-16" }` Add this.

This is the only thing that seems to work. I tried for days to fix this issue. It's very hard to understand how to change format save for c files. There was no actual answer anywhere and i had to try and retry man things before i was able to get that to work. Then trying to fix this issue. I have no idea if this is the right solution. Only it's one that seems to work.

### LunarVim version

May 3 2022

### Neovim version (>= 0.7)

NVIM v0.7.0 LuaJIT 2.1.0-beta3

### Operating system/version

Arch Linux

### Affected language servers

clang-format clangd

### Steps to reproduce

`~/.config/lvim/config.lua`

```local formatters = require "lvim.lsp.null-ls.formatters"
formatters.setup {
    {
        command = "clang-format",
        filetype = { "c", "cpp", "cs", "java" },
        extra_args = { "--style", "{IndentWidth: 4}" }
    }
}
```

### Actual behavior

spam
`warning multiple different client offset_decoding detected for buffer`

### log and support info

```shell
I don't know how to copy and past this info. But i don't see any errors. It does say there is clangd and null-ls both attatched to buf. but only clangd is autostart.
```

**Comment by abzcoding** [comment]
instead of changing the clang server settings, please either use https://www.lunarvim.org/configuration/05-autocommands.html or https://www.lunarvim.org/configuration/07-ftplugin.html#description

**Comment by JeremiahCheatham** [comment]
So i didn't understand the first link at all. I'm new to vim. But it seems like a different topic. The second link again this seems like it's affecting tab spacing that i actually already changed but doesn't effect clangd formatting that changes on save or on format back to 2 spaces. Again i don't know vim. But i have already changed the tab spacing settings.

**Comment by JeremiahCheatham** [comment]
Also if I am missing something here as i am sure i am. Can you please direct me to the actual file needed to be modified and the actual change that is needed to be made.

**Comment by abzcoding** [comment]
the first link uses autocommands ( for example on entering a c file change tab to 4 spaces)
the second link uses ftplugins ( for example to set a default value of 4 spaces for c++ files)
for clangd formatting i would suggest using `.clang-format` file to make sure your style is persistent among peers
more info -> https://clang.llvm.org/docs/ClangFormatStyleOptions.html

**Comment by JeremiahCheatham** [comment]
`~/.config/lvim/config.lua` this file
`vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "clangd"})` add this line.

`~/.config/lvim/after/ftplugin/c.lua` this file.
`"--offset-encoding=utf-16"` add this line.

is just an alternative to what i did here.
`~/.local/share/lunarvim/lvim/lua/lvim/lsp/init.lua` this file.
`capabilities.offsetEncoding = { "utf-16" } Add this.`

and i still need this for 4 spacing on format.

`~/.config/lvim/config.lua`

```formatters.setup {
    {
        command = "clang-format",
        filetype = { "c", "cpp", "cs", "java" },
        extra_args = { "--style", "{IndentWidth: 4}" }
    }
}
```
Can this `--style="{IndentWidth: 4}"`
be directly put into this file `~/.config/lvim/after/ftplugin/c.lua`?

**Comment by JeremiahCheatham** [comment]
I have been banking my head against the wall for days on this. it's a bit frustrating.

Let me start from the beginning.
I would like to use 4 spaces not 2 for c files since i am used to this and other languages also use this.

I have set lvim tab to 4 spaces `i think correctly.` but when saving it switches back to 2 spaces since that is the way clangd works.

I managed to figure out how to get both tabs and formatting to 4 spaces but with all the spamming. Both this clangd and spamming of the updated neovim are listed all over the internet with no real answers.

So i did find a solution that i post.

I tried your solution of adding that line and creating a new file. I undid all my edits. And got no spam. But i was back to were i started tab 4 spaces saving/formatting 2 spaces. So maybe it was only for the spam. I reenabled my 4 spacing format code. And that was back to 4 spaces but also all the spam was back. So that code above didn't seem to do anything. I also tried many ways of adding the 4 spacing into your code without success. So i'm still back to the method i posted first is the only working way.

**Comment by JeremiahCheatham** [comment]
Yeah i know clang has formatting. the issue is it's formatting to 2 spaces. I just want it to be 4 and globally.

I'm not sure i understand this one. -style=file. Is this global or how can i set this up?
`you can use -style=file and have ~/.config/.clang-format so all of your projects use the same one`

**Comment by abzcoding** [comment]
ok let's work on that
first things first, please comment all changes

now you have two options
1. add this to your `~/.config/lvim/config.lua`
```lua
lvim.autocommands.custom_groups = {
  -- On entering a lua file, set the tab spacing and shift width to 8
  { "BufRead", "*.c,*.cpp", "setlocal ts=4 sw=4" },
}
local formatters = require "lvim.lsp.null-ls.formatters"
formatters.setup {
    {
        command = "clang-format",
        filetype = { "c", "cpp", "cs", "java" },
        extra_args = { "--style", "file:~/.clang-format" }
    }
}
```
and then create `~/.clang-format` and fill it with
```
BasedOnStyle: LLVM
IndentWidth: 4
```
(i'll write the second one in here as well ...)

**Comment by abzcoding** [comment]
> Yeah the -style=file does work. I did all your suggestions with everything else off. But the spam is real. the new c.lua doesn't suppress it.

ok can you run `clangd --help|grep encoding` in your terminal and verify that it has `--offset-encoding`?

**Comment by abzcoding** [comment]
also please open a `.c` or `.cpp` file and run `:LspInfo` and verify that it contains `--offset-encoding=utf-16`
<img width="1558" alt="Screen Shot 2022-05-10 at 6 10 40 PM" src="https://user-images.githubusercontent.com/10992695/167642324-78ea1252-e6c1-43d5-af0f-75ede96d5390.png">

**Comment by abzcoding** [comment]
if those were as expected, then add this to your `~/.config/lvim/config.lua`
```lua
lvim.lsp.null_ls.setup.on_init = function(new_client, _)
  new_client.offset_encoding = "utf-16"
end
```

**Comment by JeremiahCheatham** [comment]
the first clangd --help gives me
--offset-encoding=<value>       - Force the offsetEncoding used for character positions. This bypasses negotiation via client capabilities

I don't see offset listed on :LspInfo

I have to run now my son has piano. I will be back in couple hours. Also remember my very first post does have a solution to the offset_encoding spam. it seems to be the only way to suppress it.

**Comment by abzcoding** [comment]
ok let's just test this one ( **backup your config somewhere** )
```shell
# remove everything
bash <(curl -s https://raw.githubusercontent.com/lunarvim/lunarvim/rolling/utils/installer/uninstall.sh)

# reinstall
bash <(curl -s https://raw.githubusercontent.com/lunarvim/lunarvim/rolling/utils/installer/uninstall.sh)

# remove current config
rm -r ~/.config/lvim

# get my temp config
git clone https://github.com/abzcoding/zvim ~/.config/lvim
```

then open `lvim` and make sure the packer stuff are done
then try opening a c or cpp file and see if the formatting works correctly or not
if everything is fine, then add back your keybindings and the rest to `~/.config/lvim/config.lua`

**Comment by JeremiahCheatham** [comment]
# switch to bash
bash

# uninstall lvim
bash <(curl -s https://raw.githubusercontent.com/lunarvim/lunarvim/rolling/utils/installer/uninstall.sh)

# remove any lvim or nvim folders.
rm -rf ~/.config/nvim
rm -rf ~/.config/lvim
rm -rf ~/.local/share/nvim
rm -rf ~/.local/share/lunarvim

# reinstall lunarvim
bash <(curl -s https://raw.githubusercontent.com/lunarvim/lunarvim/rolling/utils/installer/install.sh)

# remove current config
rm -rf ~/.config/lvim 

# get my temp config
git clone https://github.com/abzcoding/zvim ~/.config/lvim

# launch lunarvim
lvim

# blank screen that sames unamed with white line across the top. Eventually says clang installed at bottom.
# launching lvim with c file looks the same white bar across the top. no leadar key but does format on save. No spam. But doesn't look like lunarvim more like broken nvim.

**Comment by abzcoding** [comment]
run `:PackerSync` to make sure all plugins are installed
also run `:TSUpdate` to make sure treesitter is up to date

**Comment by JeremiahCheatham** [comment]
Ah yeah now it's working. So its tabbing right and formatting correctly no spam. My question is what is the difference between this and the standard config. What am i missing if anything?

**Comment by abzcoding** [comment]
basically a single flag ` "--fallback-style=google",`
which means, use `.clang-format` if you found it, use the google syntax ( 4 spaces) otherwise

**Comment by JeremiahCheatham** [comment]
I didn't delete ~/.clang-format is this still needed or part of this solution?

**Comment by abzcoding** [comment]
> I didn't delete ~/.clang-format is this still needed or part of this solution?

that is not necessary anymore

**Comment by JeremiahCheatham** [comment]
well i deleted ~/.clang-format and now it's back to 2 spaces. :-( I think it needs to be there.

**Comment by abzcoding** [comment]
> well i deleted ~/.clang-format and now it's back to 2 spaces. :-( I think it needs to be there.

oh then keep it

**Comment by JeremiahCheatham** [comment]
yeah i recreated it and it's working. So to be clear there only thing needed here was this file and nothing else?

**Comment by abzcoding** [comment]
> yeah i recreated it and it's working. So to be clear there only thing needed here was this file and nothing else?

you can test that by commenting this part
```lua
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "clangd" })

local clangd_flags = {
  "--fallback-style=google",
  "--background-index",
  "-j=12",
  "--all-scopes-completion",
  "--pch-storage=disk",
  "--clang-tidy",
  "--log=error",
  "--completion-style=detailed",
  "--header-insertion=iwyu",
  "--header-insertion-decorators",
  "--enable-config",
  "--offset-encoding=utf-16",
  "--ranking-model=heuristics",
  "--folding-ranges",
}

local clangd_bin = "clangd"

local opts = {
  cmd = { clangd_bin, unpack(clangd_flags) },
}
require("lvim.lsp.manager").setup("clangd", opts)
```

**Comment by Rz-Rz** [comment]
I'm having the same problem, I'm not intested in modifying the tab spacing, I just want to have the error message disappear. Any solution? The provided ones don't work for me. 
ie adding this:
local capabilities = vim.lsp.protocol.make_client_capabilities() Under this line 33.
capabilities.offsetEncoding = { "utf-16" }

**Comment by abzcoding** [comment]
for anyone facing this , check this out -> https://github.com/LunarVim/LunarVim/issues/3050#issuecomment-1252688272

**Comment by Rz-Rz** [comment]
> 

Could you explain how to implement this when you don't have this particular kind of config system. I only use config.lua and slapping this into that file did not work.

**Comment by abzcoding** [comment]
> Could you explain how to implement this when you don't have this particular kind of config system. I only use config.lua and slapping this into that file did not work.

probably something like this
```lua
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "clangd" })
local capabilities = require("lvim.lsp").common_capabilities()
capabilities.offsetEncoding = { "utf-16" }
local opts = {capabilities = capabilities}
require("lvim.lsp.manager").setup("clangd", opts)
```

**Comment by morphisjustfun** [comment]
> > Could you explain how to implement this when you don't have this particular kind of config system. I only use config.lua and slapping this into that file did not work.
> 
> probably something like this
> 
> ```lua
> vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "clangd" })
> local capabilities = require("lvim.lsp").common_capabilities()
> capabilities.offsetEncoding = { "utf-16" }
> local opts = {capabilities = capabilities}
> require("lvim.lsp.manager").setup("clangd", opts)
> ```

This works perfectly, Thanks!!

**Comment by JeremiahCheatham** [comment]
So I have been using lunarvim again and it was working ok. However still stuck at 2 spaces. None of the solutions above work. All off them throw errors. lunarvim still does not read .clang-format file from project directory or home or config.

this command errors because it needs 4 parameters not 2.

vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "clangd" })

I have been all over this forum searching for any solution i have missed and dived into all the links. There seems to be no solution. There is also nothing on the website or documentation. Why is it still a problem to just get 4 spaces? It makes it much easier to read the code and jump to other languages. it's also annoying to keep telling me on the bottom right that i'm stuck at 2 spaces. If it can tell me this. why can't i change it?

I am currently running the rolling branch for archlinux. I have just wiped it all and reinstalled it today since last week it seems there were changes made to this? But still no way to get it working here.

**Comment by JeremiahCheatham** [comment]
It seems vim has a settings for spaces. and then clang format has another. So spacing while writing c++ is handled by vim and when formating it's handled then by clang-format.

    ~/.config/lvim/config.lua

add the following.
```
vim.opt.shiftwidth = 4 -- the number of spaces inserted for each indentation
vim.opt.tabstop = 4 -- insert 2 spaces for a tab
```

```
clang-format -style=google -dump-config > ~/.clang-format
```
Edit lines in.

    ~/.clang-format

```
IndentWidth:    4
ObjCBlockIndentWidth:    4
```

**Comment by erichlf** [comment]
For anyone else that lands here with this warning the thing that worked for me was to add 
```
lvim.lsp.null_ls.setup.on_init = function(new_client, _)
  new_client.offset_encoding = "utf-8"
end
```
to my `config.lua`.
